common/lib/plugins/bluegreen/blue_green_status_provider.ts

```diff
@@ -0,0 +1,936 @@
+
+    if (latestInterimPhase && interimStatus.blueGreenPhase && interimStatus.blueGreenPhase.phase < latestInterimPhase.phase) {
+      this.rollback = true;
+      logger.debug(Messages.get("Bgd.rollback", this.bgdId));
+    }
+
+    if (!interimStatus.blueGreenPhase) {
+      return;
+    }
+
+    // Do not allow status moves backward (unless it's rollback).
+    // That could be caused by updating blue/green hosts delays.
+    if (!this.rollback) {
+      if (interimStatus.blueGreenPhase.phase >= this.latestStatusPhase.phase) {
+        this.latestStatusPhase = interimStatus.blueGreenPhase;
+      }
+    } else {
+      if (interimStatus.blueGreenPhase.phase < this.latestStatusPhase.phase) {
+        this.latestStatusPhase = interimStatus.blueGreenPhase;
+      }
+    }
+  }
+
+  protected updateStatusCache(): void {
+    this.pluginService.setStatus<BlueGreenStatus>(BlueGreenStatus, this.summaryStatus, this.bgdId);
+    this.storePhaseTime(this.summaryStatus.currentPhase);
+  }
+
+  protected updateCorrespondingHosts(): void {
+    this.correspondingHosts.clear();
+
+    const sourceInterimStatus: BlueGreenInterimStatus | null = this.interimStatuses[BlueGreenRole.SOURCE.value];
+    const targetInterimStatus: BlueGreenInterimStatus | null = this.interimStatuses[BlueGreenRole.TARGET.value];
+    if (sourceInterimStatus?.startTopology?.length > 0 && targetInterimStatus?.startTopology?.length > 0) {
+      const blueWriterHostInfo: HostInfo = this.getWriterHost(BlueGreenRole.SOURCE);
+      const greenWriterHostInfo: HostInfo = this.getWriterHost(BlueGreenRole.TARGET);
+      const sortedBlueReaderHostInfos: HostInfo[] | null = this.getReaderHosts(BlueGreenRole.SOURCE);
+      const sortedGreenReaderHostInfos: HostInfo[] | null = this.getReaderHosts(BlueGreenRole.TARGET);
+
+      if (blueWriterHostInfo) {
+        // greenWriterHostInfo can be null but that will be handled properly by corresponding routing.
+        this.correspondingHosts.set(blueWriterHostInfo.host, new Pair(blueWriterHostInfo, greenWriterHostInfo));
+      }
+
+      if (sortedBlueReaderHostInfos?.length > 0) {
+        if (sortedGreenReaderHostInfos?.length > 0) {
+          let greenIndex: number = 0;
+          sortedBlueReaderHostInfos.forEach((blueHostInfo) => {
+            this.correspondingHosts.set(blueHostInfo.host, new Pair(blueHostInfo, sortedGreenReaderHostInfos.at(greenIndex++)));
+            greenIndex %= sortedGreenReaderHostInfos.length;
+          });
+        } else {
+          sortedBlueReaderHostInfos.forEach((blueHostInfo) => {
+            this.correspondingHosts.set(blueHostInfo.host, new Pair(blueHostInfo, greenWriterHostInfo));
+          });
+        }
+      }
+    }
+    if (
+      sourceInterimStatus?.startTopology &&
+      sourceInterimStatus?.hostNames?.size > 0 &&
+      targetInterimStatus?.startTopology &&
+      targetInterimStatus?.hostNames?.size > 0
```

Comment:
in JDBC we don't check the startTopology of the source or target interim status, we only check that both statuses are not null and that their host names are not null or empty